### ISSUE TITLE
Terser 4.6.5 breaking js bundles

### ISSUE THREAD

**Issue by PascalBrouwers** [terser/terser#581]
**Bug report**

**Version 4.6.5**

**Complete CLI command or `minify()` options used**
```
THEME_FOLDER=('pub/static/frontend/ReachDigital/example/')

#Find deployed themes (languages) and copy the themes. 'us_US' becomes 'us_US_source'. We skip directories already having a '_source' suffix
find ${THEME_FOLDER[@]} -mindepth 1 -maxdepth 1 \( -not -name '*_source' \) -type d -execdir mv -f \{} \{}_source \;

#Find all *_source themes and use them as input for r.js. The output directories are the input directory with '_source' stripped again.
find ${THEME_FOLDER[@]} -mindepth 1 -maxdepth 1 \( -name '*_source' \) -type d -exec bash -c 'r.js -o requirejs-bundle-config.js baseUrl=\{} dir="${@%"_source"}"' bash {} \;

#Find the themes that don't have _source as suffix, assuming these are the ones that are now bundled.
BUNDLED_THEMES=$(find ${THEME_FOLDER[@]} -mindepth 1 -maxdepth 1 \( -not -name '*_source' \) -type d);

#Find all .js files in Bundled Themes (that are not already provided as minified) and run them through Terser.
find ${BUNDLED_THEMES[@]} -path '*bundles*' \( -name '*.js' -not -name '*.min.js' \) -exec terser \{} -c --comments `false` -m reserved=['$','jQuery','define','require','exports']$
find ${BUNDLED_THEMES[@]} \( -name 'require.js' -o -name 'requirejs-config.js' -o -name 'mixins.js' \) -exec terser \{} -c --comments `false` -m reserved=['$','jQuery','define','r$

# Clean up all copied *_source
find ${THEME_FOLDER[@]} -mindepth 1 -maxdepth 1 \( -name '*_source' \) -type d -exec rm -rf \{} \;
```

Hard to explain but the new version of terser breaks the js bundles that are created using this script. Everything still works fine when we use terser 4.6.4

**Comment by fabiosantoscode** [comment]
Hey there! Thanks for reporting.

I can't see the whole terser commands, as they're cut off with a `$` at the end.

Could you try to provide a small/minimal reproduction of the issue?

Also please specify how the bundles are broken. Are they trying to access variables which don't exist, or the wrong variables?

Does this issue disappear when you disable compression, mangling, or specific compress/mangle options?

**Comment by Gemorroj** [comment]
same problem.
this release breaks build angular projects with param `--prod`

**Comment by fabiosantoscode** [comment]
@Gemorroj what is breaking? Is it a crash?

**Comment by Gemorroj** [comment]
@fabiosantoscode no. project builds, but didn't work. sry, I don't have any logs at the moment.

**Comment by L2jLiga** [comment]
Angular CLI 9.0.4

```bash
npm install -g @angular/cli@latest
ng new test-proj
cd test-proj
npm install
npx ng build --prod
npm http-server
```

when open in browser:
```
main-es2015.b166bfd735935b68f734.js:1 ERROR TypeError: Cannot read property '_moduleDoBootstrap' of undefined
    at main-es2015.b166bfd735935b68f734.js:1
    at l.invoke (polyfills-es2015.072b096f4e1be3bf80df.js:1)
    at Object.onInvoke (main-es2015.b166bfd735935b68f734.js:1)
    at l.invoke (polyfills-es2015.072b096f4e1be3bf80df.js:1)
    at i.run (polyfills-es2015.072b096f4e1be3bf80df.js:1)
    at polyfills-es2015.072b096f4e1be3bf80df.js:1
    at l.invokeTask (polyfills-es2015.072b096f4e1be3bf80df.js:1)
    at Object.onInvokeTask (main-es2015.b166bfd735935b68f734.js:1)
    at l.invokeTask (polyfills-es2015.072b096f4e1be3bf80df.js:1)
    at i.runTask (polyfills-es2015.072b096f4e1be3bf80df.js:1)
```

**Comment by fabiosantoscode** [comment]
I'm publishing a hotfix revert. Please don't close this as I'll be working towards a real fix when I have the time.

Terser needs even more functional tests.

**Comment by L2jLiga** [comment]
Same without mangle

**Comment by fabiosantoscode** [comment]
4.6.6 published. Thanks for bringing this up.

**Comment by L2jLiga** [comment]
I found such repro:

*command*: `terser input.js -c passes=3 --ecma=8 -b --toplevel -o output.js`

*input*:
```javascript
const moduleRef = {
    hooks: [],
    onDestroy(cb) {
        this.hooks["onDestroy"] = cb;
    },

    onInit(cb) {
        this.hooks["onInit"] = cb;
    },
    run(callback) {
        callback();
    }
};

class Yellow {
    method() {
        return moduleRef.run(() => {
            const handler = (error) => {
                console.error("FAIL");
            };

            moduleRef.onDestroy(() => console.info("Destroyed"));
            moduleRef.onInit(() => console.info("Inited"));

            return applyCb(handler, (() => {
                return Promise.resolve().then(() => {
                    return this.message();
                });
            }));
        });
    }

    message() {
        return "PASS";
    }
}

function applyCb(handler, callback) {
    const result = callback();

    if (result.then) {
        return result.then(console.log)
            .catch(handler);
    }

    return Promise.resolve(result.message);
}

(new Yellow()).method(); // PASS
```

*output*:

```javascript
const moduleRef = {
    hooks: [],
    onDestroy(cb) {
        this.hooks.onDestroy = cb;
    },
    onInit(cb) {
        this.hooks.onInit = cb;
    },
    run(callback) {
        callback();
    }
};

(new class {
    method() {
        return moduleRef.run(() => (moduleRef.onDestroy(() => console.info("Destroyed")),
            moduleRef.onInit(() => console.info("Inited")), function(handler, callback) {
            const result = (() => Promise.resolve().then(() => this.message()))();
            return result.then ? result.then(console.log).catch(error => {
                console.error("FAIL");
            }) : Promise.resolve(result.message);
        }()));
    }
    message() {
        return "PASS";
    }
}).method(); // FAIL
```

**Comment by fabiosantoscode** [comment]
Nice repro @L2jLiga. I've pushed a fix and 4.6.5's performance boost will be back later.

**Comment by L2jLiga** [comment]
Good job!
